strawberry/relay/fields.py

```diff
@@ -0,0 +1,553 @@
+        if isinstance(nodes, Connection):
+            return nodes
+
+        return_type = cast(Connection[Node], info.return_type)
+        type_def = return_type._type_definition  # type:ignore
+        assert isinstance(type_def, TypeDefinition)
+
+        field_type = type_def.type_var_map[cast(TypeVar, NodeType)]
+        if isinstance(field_type, LazyType):
+            field_type = field_type.resolve_type()
+
+        if nodes is None:
+            nodes = cast(Node, field_type).resolve_nodes(info=info)
+
+        if inspect.isawaitable(nodes):
+
+            async def resolver():
+                return self.resolver(
+                    source,
+                    info,
+                    args,
+                    kwargs,
+                    nodes=await cast(Awaitable, nodes),
+                )
+
+            return resolver()
+
+        # Avoid info being passed twice in case the custom resolver has one
+        kwargs.pop("info", None)
+        return self.resolve_connection(cast(Iterable[Node], nodes), info, **kwargs)
+
+    def resolve_connection(
+        self,
+        nodes: Iterable[Node],
+        info: Info,
+        **kwargs,
+    ):
+        return_type = cast(Connection[Node], info.return_type)
+        return return_type.from_nodes(nodes, **kwargs)
+
+
+def node(
+    *,
+    name: Optional[str] = None,
+    is_subscription: bool = False,
+    description: Optional[str] = None,
+    permission_classes: Optional[List[Type[BasePermission]]] = None,
+    deprecation_reason: Optional[str] = None,
+    default: Any = dataclasses.MISSING,
+    default_factory: Union[Callable[..., object], object] = dataclasses.MISSING,
+    metadata: Optional[Mapping[Any, Any]] = None,
+    directives: Optional[Sequence[object]] = (),
+    # This init parameter is used by pyright to determine whether this field
+    # is added in the constructor or not. It is not used to change
+    # any behavior at the moment.
+    init: Literal[True, False, None] = None,
+) -> Any:
+    """Annotate a property to create a relay query field.
+
+    Examples:
+        Annotating something like this:
+
+        >>> @strawberry.type
+        >>> class X:
+        ...     some_node: SomeType = relay.node(description="ABC")
+
+        Will produce a query like this that returns `SomeType` given its id.
+
+        ```
+        query {
+            someNode (id: ID) {
+                id
+                ...
+            }
+        }
+        ```
+
+    """
+    return NodeField(
+        python_name=None,
+        graphql_name=name,
+        type_annotation=None,
+        description=description,
+        is_subscription=is_subscription,
+        permission_classes=permission_classes or [],
+        deprecation_reason=deprecation_reason,
+        default=default,
+        default_factory=default_factory,
+        metadata=metadata,
+        directives=directives or (),
+    )
+
+
+@overload
+def connection(
+    *,
+    resolver: Callable[[], _T],
```

Comment:
should this be resolver type?

Reply:
Oh, didn't even notice that the typing is bad in here. WIll fix it